### ISSUE TITLE
Potential data loss in DynamoDB backend

### ISSUE THREAD

**Issue by sbutler** [hashicorp/vault#5836]
**Describe the bug**
I was reviewing the code for the DynamoDB backend and noticed an improper use of the BatchWriteItem operation. The code in question is in batchWriteRequests:

https://github.com/hashicorp/vault/blob/d4367e581fe117356c4d4ac9a8c8c4e514457449/physical/dynamodb/dynamodb.go#L507-L527

The problem is that BatchWriteItem can return success but not process all the items in the request. The remaining items are returned in the UnprocessedItems response parameter. The [BatchWriteItem API documentation](https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_BatchWriteItem.html) gives an example of this happening. Notice in their example that that HTTP status code returned is 200 (success) but the response contains UnprocessedItems.

Unfortunately, the Vault DynamoDB backend discards the response and only looks at the error. In this case it would skip writing some items.

**To Reproduce**
I haven't seen this in production. However, you could theoretically trigger it by:

1. Set your table WCU to something low.
2. Create an item in kv with a long path so that it [generates many prefix items](https://github.com/hashicorp/vault/blob/d4367e581fe117356c4d4ac9a8c8c4e514457449/physical/dynamodb/dynamodb.go#L278-L292).
3. This should exceed the WCU of the table and not write all the prefix items.

**Expected behavior**
batchWriteRecords should check the response from BatchWriteItem for unprocessed items and include them in the next request. Ideally it would also implement and increasingly large backoff algorithm to make it more likely that requests will not be throttled.

**Comment by jefferai** [comment]
A PR would be welcome!

**Comment by ncsazure** [comment]
Probably it is a good time to fix this, 11 months in. Data security critical.

**Comment by jefferai** [comment]
@ncsazure just to note: we didn't write the Dynamo backend and don't support it -- our documentation is clear that it's unsupported by HashiCorp and community support only. We will be happy to review a PR but this is not something the Vault team is currently planning to address.

**Comment by sbutler** [comment]
My job hasn't given me time to work on a patch for this. And I think since I submitted it DynamoDB has added the on-demand pricing which makes this less critical for people using that backend storage.

Whoever does fix it should come up with a backoff strategy for resubmitting unprocessed items. They shouldn't just immediately try again. That makes the patch a little more difficult to do.

**Comment by cmlara** [comment]
@michelvocks @jefferai 

Anything special I need to do to make sure someone looks at this?

**Comment by AndreasSko** [comment]
We just faced this issue today when we performed a migration from Consul to DynamoDB using `vault operator migrate`. After the migration we noticed that there were around 500 secrets missing (of around 150k entries in total), so we were forced to roll back. The migration tool did not return any error, so it could have been easily missed.
We investigated the issue and it turns out that DynamoDB table was set to "on-demand mode" but still had to throttle. But normally this shouldn't be a problem if the throttling was taken care of correctly. To be honest, we are kind of shocked that this issue already exists for over two years and has not been taken with more priority. There is good chance that others have faced similar issues and might have lost secrets without even noticing.

If someone else wants to check, I created a small "hack" extending `vault operator migrate` to verify if everything has been migrated properly: https://github.com/AndreasSko/vault/commit/42d41e34fddd1fc2524d278e6186cce21b0e9871
